README.md
st7036
======

ST7036 LCD driver


packaging/CHANGELOG
st7036 (1.4.1) UNRELEASED; urgency=low

  * Added LICENSE.txt
  * Removed interpreter declaration

 -- Phil Howard <phil@pimoroni.com>  Tue, 19 Jul 2016 00:00:00 +0000

st7036 (1.4.0) UNRELEASED; urgency=low

  * Moved to Stable status
  * Added reset_pin support for Display-o-Tron HAT

 -- Phil Howard <phil@pimoroni.com>  Wed, 12 Aug 2015 00:00:00 +0000

st7036 (1.3.1) UNRELEASED; urgency=low

  * Fractional additional delay to prevent weirdness on Pi 2

 -- Phil Howard <phil@pimoroni.com>  Thu, 19 Feb 2015 00:00:00 +0000

st7036 (1.3.0) UNRELEASED; urgency=low

  * Python 3 fix

 -- Phil Howard <phil@pimoroni.com>  Mon, 03 Nov 2014 00:00:00 +0000

st7036 (1.2.0) UNRELEASED; urgency=low

  * Added "home" to reset cursor position
  * Call "home" after "clear"

 -- Phil Howard <phil@pimoroni.com>  Mon, 08 Sep 2014 12:00:00 +0000

st7036 (1.1.0) UNRELEASED; urgency=low

  * Animation/custom char support

 -- Phil Howard <phil@pimoroni.com>  Mon, 08 Sep 2014 00:00:00 +0000

st7036 (1.0.0) UNRELEASED; urgency=low

  * Initial release

 -- Jonathan Williamson <jon@pimoroni.com>  Wed, 06 Aug 2014 00:00:00 +0000


examples/test.py
#!/usr/bin/env python

import datetime
import dogbl
import threading
import time

import doglcd


lcd = doglcd.DogLCD(10,11,25,8,-1,-1)

lcd.begin(doglcd.DOG_LCD_M163, 0x28)

lcd.clear()
lcd.home()
lcd.write(chr(0) + 'OMG! Such time' + chr(0))
lcd.setCursor(0,2)
lcd.write(chr(1) + chr(4) + ' Very Wow! ' + chr(3) + chr(2) + chr(5))

bl = dogbl.DogBL(1)
bl.RGB(100,0,50)
bl.update()

class StoppableThread(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self)
        self.stop_event = threading.Event()
        self.daemon = True

    def start(self):
        if not self.isAlive():
            self.stop_event.clear()
            threading.Thread.start(self)

    def stop(self):
        if self.isAlive():
            self.stop_event.set()
            self.join()


class AsyncWorker(StoppableThread):
    def __init__(self, fn):
        StoppableThread.__init__(self)
        self.fn = fn
        self.iterations = 0

    def run(self):
        while not self.stop_event.is_set():
            if not self.fn(self.iterations):
                break
            self.iterations += 1

def dosweep(i):
    hue = (i*4)%360
    bl.sweep(hue,40)
    time.sleep(0.05)
    return True

blfade = AsyncWorker(dosweep)

try:
    blfade.start()
except KeyboardInterrupt:
    blfade.stop()
    raise

pirate = [
    [0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
    [0x00,0x1f,0x16,0x06,0x00,0x08,0x03,0x1e],
    [0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
    [0x00,0x1f,0x05,0x01,0x00,0x02,0x08,0x07]
]

heart = [
    [0x00,0x0a,0x1f,0x1f,0x1f,0x0e,0x04,0x00],
    [0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00],
    [0x00,0x00,0x00,0x0e,0x04,0x00,0x00,0x00],
    [0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00]
    ]

raa = [
    [0x1f,0x1d,0x19,0x13,0x17,0x1d,0x19,0x1f],
    [0x1f,0x17,0x1d,0x19,0x13,0x17,0x1d,0x1f],
    [0x1f,0x13,0x17,0x1d,0x19,0x13,0x17,0x1f],
    [0x1f,0x19,0x13,0x17,0x1d,0x19,0x13,0x1f]
    ]

arr = [
    [31,14,4,0,0,0,0,0],
    [0,31,14,4,0,0,0,0],
    [0,0,31,14,4,0,0,0],
    [0,0,0,31,14,4,0,0],
    [0,0,0,0,31,14,4,0],
    [0,0,0,0,0,31,14,4],
    [4,0,0,0,0,0,31,14],
    [14,4,0,0,0,0,0,31]
]

char = [
    [12,11,9,9,25,25,3,3],
    [0,15,9,9,9,25,27,3],
    [3,13,9,9,9,27,27,0],
    [0,15,9,9,9,25,27,3]
]

pacman = [
    [0x0e,0x1f,0x1d,0x1f,0x18,0x1f,0x1f,0x0e],
    [0x0e,0x1d,0x1e,0x1c,0x18,0x1c,0x1e,0x0f]
]

def getAnimFrame(char,fps):
    return char[ int(round(time.time()*fps) % len(char)) ]

while 1:
    lcd.createChar(0,getAnimFrame(char,4))
    lcd.createChar(1,getAnimFrame(arr,16))
    lcd.createChar(2,getAnimFrame(raa,8))
    lcd.createChar(3,getAnimFrame(pirate,2))
    lcd.createChar(4,getAnimFrame(heart,4))
    lcd.createChar(5,getAnimFrame(pacman,3))
    lcd.setCursor(0,1)
    t = datetime.datetime.now().strftime("%H:%M:%S.%f")
    lcd.write(t)


python/examples/test.py
#!/usr/bin/env python

import doglcd, datetime, time, dogbl, threading

lcd = doglcd.DogLCD(10,11,25,8,-1,-1)

lcd.begin(doglcd.DOG_LCD_M163, 0x28)

lcd.clear()
lcd.home()
lcd.write(chr(0) + 'OMG! Such time' + chr(0))
lcd.setCursor(0,2)
lcd.write(chr(1) + chr(4) + ' Very Wow! ' + chr(3) + chr(2) + chr(5))

bl = dogbl.DogBL(1)
bl.RGB(100,0,50)
bl.update()

class StoppableThread(threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)
		self.stop_event = threading.Event()
		self.daemon = True

	def start(self):
		if not self.isAlive():
			self.stop_event.clear()
			threading.Thread.start(self)

	
	def stop(self):
		if self.isAlive():
			self.stop_event.set()
			self.join()


class AsyncWorker(StoppableThread):
	def __init__(self, fn):
		StoppableThread.__init__(self)
		self.fn = fn
		self.iterations = 0

	def run(self):
		while not self.stop_event.is_set():
			if not self.fn(self.iterations):
				break
			self.iterations += 1

def dosweep(i):
	hue = (i*4)%360
	bl.sweep(hue,40)
	time.sleep(0.05)
	return True

blfade = AsyncWorker(dosweep)

try:
	blfade.start()
except KeyboardInterrupt:
	blfade.stop()
	raise

pirate = [
	[0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
	[0x00,0x1f,0x16,0x06,0x00,0x08,0x03,0x1e],
	[0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
	[0x00,0x1f,0x05,0x01,0x00,0x02,0x08,0x07]
]

heart = [
	[0x00,0x0a,0x1f,0x1f,0x1f,0x0e,0x04,0x00],
	[0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00],
	[0x00,0x00,0x00,0x0e,0x04,0x00,0x00,0x00],
	[0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00]
	]

raa = [
	[0x1f,0x1d,0x19,0x13,0x17,0x1d,0x19,0x1f],
	[0x1f,0x17,0x1d,0x19,0x13,0x17,0x1d,0x1f],
	[0x1f,0x13,0x17,0x1d,0x19,0x13,0x17,0x1f],
	[0x1f,0x19,0x13,0x17,0x1d,0x19,0x13,0x1f]
	]

arr = [
	[31,14,4,0,0,0,0,0],
	[0,31,14,4,0,0,0,0],
	[0,0,31,14,4,0,0,0],
	[0,0,0,31,14,4,0,0],
	[0,0,0,0,31,14,4,0],
	[0,0,0,0,0,31,14,4],
	[4,0,0,0,0,0,31,14],
	[14,4,0,0,0,0,0,31]
]

char = [
	[12,11,9,9,25,25,3,3],
	[0,15,9,9,9,25,27,3],
	[3,13,9,9,9,27,27,0],
	[0,15,9,9,9,25,27,3]
]

pacman = [
	[0x0e,0x1f,0x1d,0x1f,0x18,0x1f,0x1f,0x0e],
	[0x0e,0x1d,0x1e,0x1c,0x18,0x1c,0x1e,0x0f]
]

def getAnimFrame(char,fps):
	return char[ int(round(time.time()*fps) % len(char)) ]

while 1:
	lcd.createChar(0,getAnimFrame(char,4))
	lcd.createChar(1,getAnimFrame(arr,16))
	lcd.createChar(2,getAnimFrame(raa,8))
	lcd.createChar(3,getAnimFrame(pirate,2))
	lcd.createChar(4,getAnimFrame(heart,4))
	lcd.createChar(5,getAnimFrame(pacman,3))
	lcd.setCursor(0,1)
	t = datetime.datetime.now().strftime("%H:%M:%S.%f")
	lcd.write(t)
	


examples/sysinfo.py
import datetime
import fcntl
import math
import socket
import struct
import threading
import time

try:
    import psutil
except ImportError:
    exit("This script requires the psutil module\nInstall with: sudo pip install psutil")

import dogbl
import doglcd


def get_ip_address(ifname):
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        return socket.inet_ntoa(fcntl.ioctl(s.fileno(),
                                        0x8915, # SIOCGIFADDR
                                        struct.pack('256s', ifname[:15])
                                    )[20:24])
    except IOError:
        return ifname + ' not found!'

lcd = doglcd.DogLCD(10,11,25,8,-1,-1)

lcd.begin(doglcd.DOG_LCD_M163, 0x28)
lcd.clear()
lcd.home()
lcd.noCursor()
lcd.noDoubleHeight()
lcd.noAutoscroll()

bl = dogbl.DogBL(1)
bl.RGB(100,0,50)
bl.update()

class StoppableThread(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self)
        self.stop_event = threading.Event()
        self.daemon = True

    def start(self):
        if not self.isAlive():
            self.stop_event.clear()
            threading.Thread.start(self)


    def stop(self):
        if self.isAlive():
            self.stop_event.set()
            self.join()


class AsyncWorker(StoppableThread):
    def __init__(self, fn):
        StoppableThread.__init__(self)
        self.fn = fn
        self.iterations = 0

    def run(self):
        while not self.stop_event.is_set():
            if not self.fn(self.iterations):
                break
            self.iterations += 1

def dosweep(i):
    hue = i%360
    bl.sweep(hue,20)
    time.sleep(0.05)
    return True

blfade = AsyncWorker(dosweep)

try:
    blfade.start()
except KeyboardInterrupt:
    blfade.stop()
    raise

DOG_LCD_CHEVRON_L = 0b11111011
DOG_LCD_CHEVRON_R = 0b11111100

# We need some custom characters to
# complete the levels of bar graph
lcd.createChar(6,[0,31,0,0,0,0,0,0])
lcd.createChar(5,[0,0,31,0,0,0,0,0])
lcd.createChar(4,[0,0,0,0,31,0,0,0])
lcd.createChar(3,[0,0,0,0,0,31,0,0])
lcd.createChar(2,[0,0,0,0,0,0,0,31])

# Always call home after creating
# custom chars
lcd.home()

# Map out the levels of bar graph
graph_steps = [2,95,3,4,45,5,6,255]

graph_text = []

while 1:
    #lcd.setCursor(0,0)
    #lcd.write(get_ip_address('eth0'))

    #lcd.setCursor(0,1)
    #lcd.write(get_ip_address('wlan0'))

    ip = psutil.net_connections(kind='inet4')
    mem = psutil.virtual_memory()
    cpu = psutil.cpu_percent()

    lcd.setCursor(0,0)
    lcd.write("FREE " + str(mem.available/1024/1024) + 'MB/' + str(mem.total/1024/1024) + 'MB')

    lcd.setCursor(0,1)
    lcd.write("CPU% " + str(round(cpu,2)) + '    ')

    idx = int(math.floor((7.0/100.0) * cpu))

    graph_text.append(graph_steps[idx])

    if len(graph_text) > 16:
        graph_text.pop(0)

    lcd.setCursor(0,2)
    for char in graph_text:
        lcd.writeChar(char)

    time.sleep(0.25)


library/st7036.py
import time

try:
    import spidev
except ImportError:
    exit("This library requires the spidev module\nInstall with: sudo pip install spidev")

try:
    import RPi.GPIO as GPIO
except ImportError:
    exit("This library requires the RPi.GPIO module\nInstall with: sudo pip install RPi.GPIO")


COMMAND_CLEAR = 0b00000001
COMMAND_HOME = 0b00000010
COMMAND_SCROLL = 0b00010000
COMMAND_DOUBLE = 0b00010000
COMMAND_BIAS = 0b00010100
COMMAND_SET_DISPLAY_MODE = 0b00001000

BLINK_ON = 0b00000001
CURSOR_ON = 0b00000010
DISPLAY_ON = 0b00000100
TOP = 1
BOTTOM = 0

class st7036():
    def __init__(self,
                 register_select_pin,
                 reset_pin=None,
                 rows=3,
                 columns=16,
                 spi_chip_select=0,
                 instruction_set_template=0b00111000):

        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)

        self.spi = spidev.SpiDev()
        self.spi.open(0, spi_chip_select)
        self.spi.max_speed_hz = 1000000

        self.reset_pin = reset_pin
        self.row_offsets = ([0x00], [0x00, 0x40], [0x00, 0x10, 0x20])[rows - 1]
        self.rows = rows
        self.columns = columns

        if self.reset_pin is not None:
            GPIO.setup(self.reset_pin,  GPIO.OUT)
            GPIO.output(self.reset_pin, GPIO.LOW)
            time.sleep(0.001)
            GPIO.output(self.reset_pin, GPIO.HIGH)
            time.sleep(0.001)

        GPIO.setup(register_select_pin, GPIO.OUT)
        GPIO.output(register_select_pin, GPIO.HIGH)

        self.register_select_pin = register_select_pin
        self.instruction_set_template = instruction_set_template

        self._enabled = True
        self._cursor_enabled = False
        self._cursor_blink = False
        self._double_height = 0

        self.animations = [None]*8

        self.update_display_mode()

        # set entry mode (no shift, cursor direction)
        self._write_command(0b00000100 | 0b00000010)

        self.set_bias(1)

        self.set_contrast(40)
        self.clear()

    def reset(self):
        if self.reset_pin is not None:
            GPIO.output(self.reset_pin, GPIO.LOW)
            time.sleep(0.001)
            GPIO.output(self.reset_pin, GPIO.HIGH)

    def set_bias(self, bias=1):
        self._write_command(COMMAND_BIAS | (bias << 4) | 1, 1)

    def set_contrast(self, contrast):
        """
        Sets the display contrast.

        Args:
            contrast (int): contrast value
        Raises:
            TypeError: if contrast is not an int
            ValueError: if contrast is not in the range 0..0x3F
        """
        if type(contrast) is not int:
            raise TypeError("contrast must be an integer")

        if contrast not in range(0, 0x40):
            raise ValueError("contrast must be an integer in the range 0..0x3F")

        # For 3.3v operation the booster must be on, which is
        # on the same command as the (2-bit) high-nibble of contrast
        self._write_command((0b01010100 | ((contrast >> 4) & 0x03)), 1)
        self._write_command(0b01101011, 1)

        # Set low-nibble of the contrast
        self._write_command((0b01110000 | (contrast & 0x0F)), 1)

    def set_display_mode(self, enable=True, cursor=False, blink=False):
        """
        Sets the display mode.

        Args:
            enable (boolean): enable display output
            cursor (boolean): show cursor
            blink (boolean): blink cursor (if shown)
        """
        self._enabled = enable
        self._cursor_enabled = cursor
        self._cursor_blink = blink
        self.update_display_mode()

    def update_display_mode(self):
        mask = COMMAND_SET_DISPLAY_MODE
        mask |= DISPLAY_ON if self._enabled else 0
        mask |= CURSOR_ON if self._cursor_enabled else 0
        mask |= BLINK_ON if self._cursor_blink else 0
        self._write_command(mask)

    def enable_cursor(self, cursor=False):
        self._cursor_enabled = cursor
        self.update_display_mode()

    def enable_blink(self, blink=False):
        self._cursor_blink = blink
        self.update_display_mode()

    def set_cursor_offset(self, offset):
        """
        Sets the cursor position in DRAM

        Args:
            offset (int): DRAM offset to place cursor
        """
        self._write_command(0b10000000 | offset)

    def set_cursor_position(self, column, row):
        """
        Sets the cursor position in DRAM based on
        a row and column offset

        Args:
            column (int): column to move the cursor to
            row (int): row to move the cursor to
       Raises:
            ValueError: if row and column are not within defined screen size
        """
        if row not in range(self.rows) or column not in range(self.columns):
            raise ValueError("row and column must integers within the defined screen size")

        offset = self.row_offsets[row] + column

        self._write_command(0b10000000 | offset)

    def home(self):
        """
        Sets the cursor position to 0,0
        """
        self.set_cursor_position(0, 0)

    def clear(self):
        """
        Clears the display and resets the cursor.
        """
        self._write_command(COMMAND_CLEAR)
        self.home()

    def write(self, value):
        """
        Write a string to the current cursor position.

        Args:
            value (string): The string to write
        """
        GPIO.output(self.register_select_pin, GPIO.HIGH)

        for i in [ord(char) for char in value]:
            self.spi.xfer([i])
            time.sleep(0.00005)

    def create_animation(self, anim_pos, anim_map, frame_rate):
        """Creates an animation in a given custom character slot

        Args:
            anim_pos (int): Character slot from 0 to 7 where current animation frame should be store on LCD
            anim_map: A list of custom character definitions. Each definition should be a list of 8 bytes describing the custom character for that frame,
            frame_rate: Speed of animation in frames-per-second
        """
        if anim_pos >= len(self.animations):
            raise ValueError("Valid animation positions are 0 to 7")

        if type(anim_map) is not list:
            raise ValueError("Animation map should be a list of animation frames")

        if type(anim_map[0]) is not list or len(anim_map[0]) < 8:
            raise ValueError("Animation frames should be lists of 8 bytes")

        self.create_char(anim_pos, anim_map[0])
        self.animations[anim_pos] = [anim_map, frame_rate]
        self.set_cursor_position(0, 1)

    def update_animations(self):
        for i, animation in enumerate(self.animations):
            if animation is not None and len(animation) == 2:
                anim = animation[0]
                fps = animation[1]
                frame = anim[int(round(time.time()*fps) % len(anim))]
                self.create_char(i, frame)
        self.set_cursor_position(0, 1)

    def create_char(self, char_pos, char_map):
        if char_pos < 0 or char_pos > 7:
            return False

        baseAddress = char_pos*8
        for i in range(0, 8):
            self._write_command((0x40 | (baseAddress+i)))
            self._write_char(char_map[i])

        self.set_display_mode()

    def cursor_left(self):
        self._write_command(COMMAND_SCROLL, 0)

    def cursor_right(self):
        self._write_command(COMMAND_SCROLL | (1 << 2), 0)

    def shift_left(self):
        self._write_command(COMMAND_SCROLL | (1 << 3), 0) # 0x18

    def shift_right(self):
        self._write_command(COMMAND_SCROLL | (1 << 3) | (1 << 2), 0) # 0x1C

    def double_height(self, enable=0, position=1):
        self._double_height = enable
        self._write_instruction_set(0)
        self._write_command(COMMAND_DOUBLE | (position << 3), 2)

    def _write_char(self, value):
        GPIO.output(self.register_select_pin, GPIO.HIGH)
        self.spi.xfer([value])

        time.sleep(0.0001)

    def _write_instruction_set(self, instruction_set=0):
        GPIO.output(self.register_select_pin, GPIO.LOW)
        self.spi.xfer([self.instruction_set_template | instruction_set | (self._double_height << 2)])
        time.sleep(0.00006)

    def _write_command(self, value, instruction_set=0):
        GPIO.output(self.register_select_pin, GPIO.LOW)

        # select correct instruction set
        self._write_instruction_set(instruction_set)

        # switch to command-mode
        self.spi.xfer([value])

        time.sleep(0.00006)

if __name__ == "__main__":
    print("st7036 test cycles")

    import time
    import sys
    import os
    import random

    # disable output buffering for our test activity dots
    sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)

    lcd = st7036(register_select_pin=25, rows=3, columns=16)
    lcd.set_display_mode()
    lcd.set_contrast(40)
    lcd.clear()

    print(">> fill screen")
    for i in range(48):
        lcd.set_cursor_offset(i)
        time.sleep(.05)
        lcd.write(chr(i+65))
        time.sleep(.02)

    print(">> cycle character set")
    for i in range(256 - 48 - 65):
        lcd.set_cursor_offset(0x00)
        lcd.write("".join([chr(i + j + 65) for j in range(48)]))
        time.sleep(.02)
        lcd.clear()
    lcd.clear()

    print(">> test contrast range")
    lcd.set_cursor_offset(0x10)
    lcd.write("test contrast")
    for i in range(0x40):
        lcd.set_contrast(i)
        time.sleep(0.02)
    for i in reversed(range(0x40)):
        lcd.set_contrast(i)
        time.sleep(0.02)
        
    lcd.set_contrast(40)
    lcd.clear()

    print(">> test set cursor position")
    for i in range(50):
        row = random.randint(0, 3 - 1)
        column = random.randint(0, 16 - 1)

        lcd.set_cursor_position(column, row)
        lcd.write(chr(0b01101111))
        time.sleep(.10)
        lcd.set_cursor_position(column, row)
        lcd.write(" ")
